evm/vm/forks/sharding/contracts/validator_manager.v.py

```diff
@@ -237,37 +251,68 @@ def add_header(
         receipt_root,
         convert(collation_number, 'bytes32'),
     )
-    entire_header_hash: bytes32 = sha3(header_bytes)
-    assert self.collation_headers[shard_id][entire_header_hash].score == 0
-    # Check whether the parent exists.
-    # if (parent_hash == 0), i.e., is the genesis,
-    # then there is no need to check.
-    if parent_hash != convert(0, 'bytes32'):
-        assert self.collation_headers[shard_id][parent_hash].score > 0
+    entire_header_hash: bytes32 = convert(
+        uint256_mod(
+            convert(sha3(header_bytes), 'uint256'),
+            # Mod 2^192, i.e., extract right most 24 bytes
+            convert(6277101735386680763835789423207666416102355444464034512896, 'uint256')
+        ),
+        'bytes32'
+    )
+    # Check that this header has not yet been added.
+    assert not self.collation_headers[shard_id][entire_header_hash]
```

Comment:
Just saying, if you move `assert self.period_head[shard_id] < expected_period_number` in front, then you don't need to check `assert not self.collation_headers[shard_id][entire_header_hash]`.... unless SHA-3 collision(!)

Reply:
Good point. Though if we implement slashing condition for adding multiple header at same shard during same period then `self.period_head` can be removed.